src/protocol/basics/share_known_value/mod.rs

```diff
@@ -0,0 +1,41 @@
+use crate::bits::Fp2Array;
+use crate::ff::Field;
+use crate::helpers::Role;
+use crate::protocol::context::{Context, SemiHonestContext, MaliciousContext};
+use crate::secret_sharing::SharedValue;
+use crate::secret_sharing::replicated::semi_honest::{AdditiveShare as Replicated, XorShare};
+use crate::secret_sharing::replicated::malicious::AdditiveShare as MaliciousReplicated;
+
+pub trait ShareKnownValue<C: Context, V: SharedValue> {
+    fn share_known_value(ctx: &C, value: V) -> Self;
+}
+
+impl<'a, F: Field> ShareKnownValue<SemiHonestContext<'a>, F> for Replicated<F> {
+    fn share_known_value(ctx: &SemiHonestContext<'a>, value: F) -> Self {
+        match ctx.role() {
+            Role::H1 => Self::new(value, F::ZERO),
+            Role::H2 => Self::new(F::ZERO, F::ZERO),
+            Role::H3 => Self::new(F::ZERO, value),
+        }
+    }
+}
+
+impl<'a, V: Fp2Array> ShareKnownValue<SemiHonestContext<'a>, V> for XorShare<V> {
+    fn share_known_value(ctx: &SemiHonestContext<'a>, value: V) -> Self {
+        match ctx.role() {
+            Role::H1 => Self::new(value, V::ZERO),
+            Role::H2 => Self::new(V::ZERO, V::ZERO),
+            Role::H3 => Self::new(V::ZERO, value),
+        }
+    }
+}
+
+impl<'a, F: Field> ShareKnownValue<MaliciousContext<'a, F>, F> for MaliciousReplicated<F> {
+    fn share_known_value(ctx: &MaliciousContext<'a, F>, value: F) -> Self {
+        // TODO: I don't actually think r_share should be exposed like this
```

Comment:
Say more?  You are probably right, in that the share should not be used, except under tightly controlled conditions, but this seems like pretty clean code and an exception to the usual rule where values should be upgraded and added to the MAC, rather than using r_share().

Reply:
The thing that made me cringe (and where I first added this comment) was what this calls -- `pub fn r_share(&self)` on MaliciousContext. To me it's a similar argument to why `MaliciousReplicated` shouldn't allow direct access to `x`.

In my cleanups I changed this so r_share is not exposed -- I moved the `known * r` calculation back inside a method on MaliciousContext.

This specific issue aside, I do have some concerns about the malicious share-known-value protocol -- it seems dangerous to me to combine the private shares of `r` with publicly known constants. The resultant shares are effectively as sensitive as shares of `r`. If you do something like share-known-value followed by reveal (admittedly a pointless combination), then you've leaked the `r` share to another party. I made a note to think about this some more -- or track down a reference for the malicious share-known-value protocol.

Reply:
Sounds like you have it in hand, which is good.

Yes, `known * r` is a problem if revealed, but that is true for any value that might be revealed.  Revealing is something we only do once we have confidence that the value is safe to reveal.

We require that `r` is revealed before revealing anything that includes it.  I wonder if there is any way to structure the code to enforce that constraint.  It's the opposite of the usual lifetime thing in Rust though (`'x: 'share - 'r` isn't a thing you can do, and probably for good reason.)